### PR TITLE
DAOS-6255 placement: few fixes for placement

### PR DESCRIPTION
**Pull request by wangdi1**
1. st_shard in shard_tgt should be the real offset
of the shard in object shards array.

2. Do not add new remap shards to the extend list for
remap shards for reintegration layout.

3. Using real group size, instead of object class group
size to calculate the group index.

4 add tests to verify large stripe rebuild/reintegration.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

File: src/placement/jump_map.c [hunk 1]
```diff
@@ -519,12 +519,6 @@ obj_remap_shards(struct pl_jump_map *jmap, struct daos_obj_md *md,
 			get_target(root, &spare_tgt, crc(key, rebuild_key),
 				   dom_used, tgts_used, shard_id, op_type);
 			spares_left--;
-			if (can_extend(op_type, spare_tgt->ta_comp.co_status)) {
-				rc = remap_alloc_one(extend_list, shard_id,
-						spare_tgt, true);
-				if (rc)
-					return rc;
-			}
```
Comment (byron-marohn) [inline src/placement/jump_map.c:527]: We don't currently have any tests that cover this, but this seems like it was probably added here for a reason.... I suspect removing this code is going to break addition somehow. Why was this removed?

Comment (wangdi1) [reply src/placement/jump_map.c:527]: For example, for shard A, the rebuild sequence with the JCH are

tgt22 ---> tgt17--->tgt25  (suppose tgt22 and tgt 17 are all in rank 2). i.e. if we kill rank2,  tgt22 will be rebuilt on tgt25 anyway, because tgt17 is also failed.

Then reintegrate rank2,  it needs move data back directly from tgt25 --> tgt22 instead of through tgt17.

During the remap,  initially it has tgt22 then it get tgt17 due to JCH, but tgt17 is also in UP status, so it will go further to find out tgt25.  But we should not add tgt 17 to the extending list,  because there won't be any data to be reintegrated on tgt17 and also this will break the object layout with inflight IO, and causing that assertion.

I assume the initial implementation is just trying follow the rebuild remap list, though this is not quite right for extending list.

**Comment by wangdi1** [comment]
reviewers,  this is required for 1.1.3, please review it asap. Thank you!

**Review by byron-marohn — APPROVED** [review]
I'd be a lot happier with this change if we had any addition tests... but priorities being what they are, this seems reasonable to me.

File: src/placement/jump_map.c (hunk 1 above)
Comment (liuxuezhao) [reply src/placement/jump_map.c:527]: in above example, tgt22 ---> tgt17--->tgt25 is the spare tgt selection list? why the two tgts (22 and 17) on same node can be adjacent to each other, in the first place?

**Review by liuxuezhao — APPROVED** [review]

**Comment by kprantis** [comment]
I see 2 approves - anything left to do on this patch prior to gatekeeping?

**Comment by wangdi1** [comment]
> I'd be a lot happier with this change if we had any addition tests... but priorities being what they are, this seems reasonable to me.

You mean addition daos core tests? yes, I will add them later.

File: src/object/srv_obj_migrate.c [hunk 2]
```diff
@@ -1897,6 +1897,13 @@ migrate_one_epoch_object(daos_handle_t oh, daos_epoch_range_t *epr,
 			/* DER_DATA_LOSS means it can not find any replicas
 			 * to rebuild the data, see obj_list_common.
 			 */
+			if (rc == -DER_DATA_LOSS) {
+				D_DEBUG(DB_REBUILD, "No replicas for "DF_UOID
+					"\n", DP_UOID(arg->oid));
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/object/srv_obj_migrate.c:1902]: Hmm, not sure, since this is not real "error". If we use D_ERROR, it might scare users, and also generate a lot of them, if there is a lot single replica object.

**Comment by wangdi1** [comment]
in above example, tgt22 ---> tgt17--->tgt25 is the spare tgt selection list? why the two tgts (22 and 17) on same node can be adjacent to each other, in the first place?

oh, we use 2G8 in a 4 server (8 targets each server) environment, so those shards are co-located anyway.

File: src/container/srv_container.c [hunk 3]
```diff
@@ -1220,16 +1220,20 @@
...
-			D_DEBUG(DB_MD, DF_CONT" sync "DF_U64"\n",
-				DP_CONT(svc->cs_pool_uuid,
-					ec_agg->ea_cont_uuid), min_eph);
+			/**
+			 * NB: during extending or reintegration, the new
+			 * server might cause the minum epoch is less than
+			 * ea_current_eph.
+			 */
+			D_DEBUG(DB_MD, DF_CONT" minum "DF_U64" cur "DF_U64"\n",
+			       DP_CONT(svc->cs_pool_uuid, ec_agg->ea_cont_uuid),
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/container/srv_container.c]: This PR landed with the warning unresolved so every PR is going to get notified about it now.

[the remark this answers is not in the document]
Comment (wangdi1) [reply src/container/srv_container.c]: oh,  I resolved this by PR-4568. Hope to land soon. Sorry.
